Enable {Mutating,Validating}AdmissionWebhook for k8s 1.9+

alvaroaleman opened the pull request:
**What this PR does / why we need it**:

Enable admission webhooks so applicants have a chance to get their exercises working :)

I would btw. be very, very much in favor of moving all non-optional fields in the kubermaticv1.Cluster struct that currently are a pointer to a literal, because:

* This makes writing tests much easier
* This reduces the chance of getting NPEs inside the controlller

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
```release-note
Enabled Mutating/Validating Admission Webhooks for K8S 1.9+
```

alvaroaleman committed:
Enable {Mutating,Validating}AdmissiobWebhook for k8s 1.9+

mrIncompetent commented on api/pkg/resources/apiserver/deployment.go:
```diff
@@ -252,6 +254,12 @@ func getApiserverFlags(data *resources.TemplateData, externalNodePort int32, etc
 	if nodePortRange == "" {
 		nodePortRange = defaultNodePortRange
 	}
+	admissionControlFlagValue := "NamespaceLifecycle,LimitRanger,ServiceAccount,DefaultStorageClass,ResourceQuota,NodeRestriction"
```
Are there constants in the `k8s` vendors for those?

mrIncompetent replied:
If so:
Can we use a String-set for this? `k8s.io/apimachinery/pkg/util/sets/string.go:27`
Join via
```go
	s := sets.NewString("foo", "bar")
	strings.Join(s.List(), ",")
```

alvaroaleman replied:
Does not look like:: `ack 'NodeRestriction' vendor/`

`MutatingAdmissionWebhook,ValidatingAdmissionWebhook`  are beta in 1.9, thats why I used 1.9 as starting point.

Ref: https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#is-there-a-recommended-set-of-admission-controllers-to-use

It is actually recommended to use `ResourceQuota`, I could add that.

Also its recommended to use a different flag, `--enable-admission-plugins` instead of `--admission-control`. I'll update the pull.

mrIncompetent replied:
Yep, we should stick to the recommendations

alvaroaleman committed:
Simplyfy flag test, use new flagname in 1.10+

alvaroaleman commented:
I've updated the unit test to just use a set to test for the expected flags and also adjusted what we have to upstream recommendations. The only difference so far is that we additionally add the `NodeRestriction`.

Using a set inside the `getApiserverFlags` is IMHO not a good idea, as the order of the flags matter pre 1.10

alvaroaleman commented:
Regarding

> I would btw. be very, very much in favor of moving all non-optional fields in the kubermaticv1.Cluster struct that currently are a pointer to a literal

Does that sound reasonable to you? If yes, I'd like to open an issue for that

mrIncompetent commented:
> Does that sound reasonable to you? If yes, I'd like to open an issue for that

:+1:

mrIncompetent commented on api/pkg/resources/apiserver/deployment.go:
```diff
@@ -252,6 +254,21 @@ func getApiserverFlags(data *resources.TemplateData, externalNodePort int32, etc
+
+	// Order of these flags matter pre 1.10 and MutatingAdmissionWebhook may manipulate the object, so "ResourceQuota
+	// must come last
+	admissionControlFlagValue += ",ResourceQuota"
```
As the logic now grew a bit, can we move it to a function?

alvaroaleman replied:
@mrIncompetent Done

alvaroaleman committed:
Use a dedicated func to for apiservers admisisonControlFlags

mrIncompetent commented on api/pkg/resources/apiserver/deployment.go:
```diff
@@ -305,6 +309,33 @@ func getApiserverFlags(data *resources.TemplateData, externalNodePort int32, etc
+	// must come last
+	admissionControlFlagValue += ",ResourceQuota"
+
+	return [2]string{admissionControlFlagName, admissionControlFlagValue}
```
Instead of returning a string slice, return 2 strings. Its more readable

alvaroaleman committed:
Return two strings instead of string array in getAdmissionControlFlags

alvaroaleman commented:
> Instead of returning a string slice, return 2 strings. Its more readable
Good point, done
